### PR TITLE
Prompt user of running workers/tasks on migration

### PR DESCRIPTION
**Pull request by werwty**
Add close() method to celery scheduler for clean exit.
Catch ServerSelectionTimeoutError when mongo is down.

closes #2186
https://pulp.plan.io/issues/2186

### CODE REVIEW

**Comment by bmbouter** [comment]
@werwty FYI there is another PR which moves code you are editing. We need to somehow coordinate these two efforts.

File: server/pulp/server/async/scheduler.py [hunk 1]
```diff
@@ -368,3 +369,6 @@ def add(self, **kwargs):
         entries to the database, and they will be picked up automatically.
         """
         raise NotImplementedError
+
+    def close(self):
```
Comment (bmbouter) [inline server/pulp/server/async/scheduler.py]: What calls this?

File: server/pulp/server/db/manage.py [hunk 2]
```diff
@@ -191,6 +193,14 @@ def main():
         options = parse_args()
         _start_logging()
         connection.initialize(max_timeout=1)
+
+        # Prompt the user if there are active workers or running tasks
+        if status.get_workers() or filter(lambda x: x.state in constants.CALL_INCOMPLETE_STATES,
```
Comment (bmbouter) [inline server/pulp/server/db/manage.py]: I think we should disinclude the check for tasks in the incomplete states. The only unsafe aspect really is if workers are running. Sometimes bugs put task states incorrectly in an incomplete state so I don't want those bugs to have an impact on this functionality.

File: server/pulp/server/db/manage.py [hunk 3]
```diff
@@ -191,6 +193,14 @@ def main():
...
+                                          model.TaskStatus.objects(group_id=None)):
+
+            if not _user_input_continue('There are still running workers/tasks,'
```
Comment (bmbouter) [inline server/pulp/server/db/manage.py]: Also what if we don't give the user an option to continue at all. It seems unsafe to continue. If we do continue to prompt we need to reword this to clearly state that it is likely unsafe to continue.

File: server/pulp/server/db/manage.py (hunk 2 above)
Comment (bmbouter) [inline server/pulp/server/db/manage.py]: Also what about checking the timestamps? This will only work if all processes are gracefully shut down. If we check the timestamps also we could pick up cases for non-graceful shutdown. What do you think?

File: server/pulp/server/db/manage.py (hunk 3 above)
Comment (bmbouter) [inline server/pulp/server/db/manage.py]: These user facing strings should be i8ln

File: server/pulp/server/async/scheduler.py (hunk 1 above)
Comment (werwty) [inline server/pulp/server/async/scheduler.py]: scheduler close is called by the celery beat service object when it does a sync:
https://github.com/celery/celery/blob/master/celery/beat.py#L544

By default scheduler.close just calls scheduler.sync, which does not seem to do much. Overriding it to clean up our db seems like the cleanest solution.
https://github.com/celery/celery/blob/master/celery/beat.py#L331

File: server/pulp/server/db/manage.py [hunk 4]
```diff
@@ -200,6 +210,9 @@ def main():
         return os.EX_DATAERR
     except models.MigrationRemovedError:
         return os.EX_SOFTWARE
+    except ServerSelectionTimeoutError:
```
Comment (werwty) [inline server/pulp/server/db/manage.py]: If the mongodb service is down the exception is raised when _CONNECTION.server_info() is called at:
https://github.com/pulp/pulp/blob/master/server/pulp/server/db/connection.py#L119

currently we're not catching this exception and show a stack trace to the user

File: server/pulp/server/db/manage.py (hunk 2 above)
Comment (werwty) [inline server/pulp/server/db/manage.py]: Good point, I'll add a worker timestamp checking and remove the task checking

File: server/pulp/server/db/manage.py (hunk 3 above)
Comment (werwty) [inline server/pulp/server/db/manage.py]: I would rather give the user an option to continue, since personally, I prefer software to trust me.
Instead I will reword this to induce more fear.

File: server/pulp/server/async/scheduler.py (hunk 1 above)
Comment (bmbouter) [inline server/pulp/server/async/scheduler.py]: We call `super()` here. It may not do much but since we are overriding it and we didn't before we should preserve to codepath.

File: server/pulp/server/db/manage.py (hunk 4 above)
Comment (bmbouter) [inline server/pulp/server/db/manage.py]: Niiiice.  :+1:

File: server/pulp/server/db/manage.py [hunk 5]
```diff
@@ -191,6 +195,15 @@ def main():
         options = parse_args()
         _start_logging()
         connection.initialize(max_timeout=1)
+
+        # Prompt the user if there are workers that have not timed out
+        if filter(lambda worker: (UTCDateTimeField().to_python(datetime.now()) -
+                                  worker['last_heartbeat']).total_seconds() <
+                  constants.CELERY_TIMEOUT_SECONDS, status.get_workers()):
+            if not _user_input_continue('There are still running workers, continuing can '
+                                        'cause major data issues. Are you sure you wish '
```
Comment (bmbouter) [inline server/pulp/server/db/manage.py]: I like the language update. I was thinking about how to introduce an `corrupt` as an even scarier word. Consider s/can cause major data issues/could corrupt your Pulp installation/  I'm OK with it as is, so this is not a required change.

File: server/pulp/server/db/manage.py [hunk 6]
```diff
@@ -191,6 +195,15 @@ def main():
         options = parse_args()
         _start_logging()
         connection.initialize(max_timeout=1)
+
+        # Prompt the user if there are workers that have not timed out
+        if filter(lambda worker: (UTCDateTimeField().to_python(datetime.now()) -
+                                  worker['last_heartbeat']).total_seconds() <
+                  constants.CELERY_TIMEOUT_SECONDS, status.get_workers()):
+            if not _user_input_continue('There are still running workers, continuing can '
```
Comment (bmbouter) [inline server/pulp/server/db/manage.py]: Can this string be i8ln

File: server/pulp/server/db/manage.py [hunk 7]
```diff
@@ -274,3 +290,13 @@ def _start_logging():
     console_handler = logging.StreamHandler()
     console_handler.setLevel(logging.INFO)
     _logger.root.addHandler(console_handler)
+
+
+def _user_input_continue(question):
+    reply = str(raw_input(_(question + ' (y/n): '))).lower().strip()
```
Comment (bmbouter) [inline server/pulp/server/db/manage.py]: Can this default to no? I think the prompt would become `(y/N`)` What do you think about this?

**Comment by bmbouter** [comment]
A release note needs to be added for this feature

File: docs/user-guide/release-notes/2.11.x.rst [hunk 8]
```diff
@@ -10,3 +10,5 @@ New Features
 
 * For RPM content, a full sync will be forced if the sync configuration has been changed or content
   has been removed since the last sync.
+
+* When pulp-manage-db is called, prompt the user to continue if there are running workers.
```
Comment (bmbouter) [inline docs/user-guide/release-notes/2.11.x.rst]: s/called/run/
s/running workers/Pulp services still running/
